src/ir/effects.h

```diff
@@ -751,10 +751,32 @@ class EffectAnalyzer {
       // traps when ref is null.
       parent.implicitTrap = true;
     }
-    void visitStringViewAccess(StringViewAccess* curr) {
+    void visitStringWTF8Advance(StringWTF8Advance* curr) {
       // traps when ref is null.
       parent.implicitTrap = true;
     }
+    void visitStringWTF16Get(StringWTF16Get* curr) {
+      // traps when ref is null.
+      parent.implicitTrap = true;
+    }
+    void visitStringIterNext(StringIterNext* curr) {
+      // traps when ref is null.
+      parent.implicitTrap = true;
+      // modifies state in the iterator. we model that as accessing heap memory
+      // in an array atm TODO consider adding a new effect type for this (we
+      // added one for arrays because struct/array operations often interleave,
+      // say with vtable accesses, but it's not clear adding overhead to this
+      // class is worth it for string iters)
+      parent.readsArray = true;
+      parent.writesArray = true;
+    }
+    void visitStringIterMove(StringIterMove* curr) {
+      // traps when ref is null.
+      parent.implicitTrap = true;
+      // see StringIterNext.
+      parent.readsArray = true;
+      parent.writesArray = true;
+    }
```

Comment:
If they have the same side effects, can't we merge `next` with `move`, with an opcode, and  if the opcode is `next`, artificially set `num` to `(i32.const 1)`? We shouldn't print the `(i32.const 1)` when emitting text or binary, but having it a real `Expression` can spare us of many `if`-`else`es, like, `if (op == move) do something; else do something else`

Reply:
Hmm, I don't think we have an example where a child exists but isn't printed in some cases? Maybe I forgot something.

Another option is for `num` to be `nullptr` for `next`. But I'm not sure how much I like that either.

Maybe we can revisit this when we get to optimizations? If the if-else pattern you mention becomes a problem then we can refactor.

Reply:
> Hmm, I don't think we have an example where a child exists but isn't printed in some cases? Maybe I forgot something.

Yeah this would be an exception.. But we already do many tweaks in wast printer (for example, omitting block printing), so I thought that wouldn't be too big an issue.

> Another option is for `num` to be `nullptr` for `next`. But I'm not sure how much I like that either.

That's an option too. I thought having if-else only when printing would be better than having it in all other computational places... Maybe I'm too optimistic?

> Maybe we can revisit this when we get to optimizations? If the if-else pattern you mention becomes a problem then we can refactor.

Yeah that sounds fine too.

Reply:
Ok, cool, let's land this for now. I agree that the wast printer workaround would be ok, so it really depends on how the optimizer code ends up looking. We'll see when we get there I guess...